### PR TITLE
[Speech Examples] Add pytorch speech pretraining

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

This PR adds a Wav2Vec2 pretraining example for PyTorch. Some experiments are still running, but it would be great if the PR could already be reviewed.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py [hunk 1]
```diff
@@ -1001,10 +1076,10 @@ def _mask_hidden_states(
                 (batch_size, sequence_length),
                 mask_prob=self.config.mask_time_prob,
                 mask_length=self.config.mask_time_length,
-                device=hidden_states.device,
+                device=self.device,
                 attention_mask=attention_mask,
-                min_masks=2,
             )
+            mask_time_indices = torch.from_numpy(mask_time_indices, device=hidden_states.device)
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/modeling_wav2vec2.py:1082]:
```suggestion
            mask_time_indices = torch.tensor(mask_time_indices, device=hidden_states.device, dtype=torch.long)
```

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py [hunk 2]
```diff
@@ -1013,10 +1088,11 @@ def _mask_hidden_states(
                 (batch_size, hidden_size),
                 mask_prob=self.config.mask_feature_prob,
                 mask_length=self.config.mask_feature_length,
-                device=hidden_states.device,
+                device=self.device,
                 attention_mask=attention_mask,
             )
-            hidden_states[mask_feature_indices[:, None].expand(-1, sequence_length, -1)] = 0
+            mask_feature_indices = torch.from_numpy(mask_feature_indices, device=hidden_states.device)[:, None].expand(-1, sequence_length, -1)
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/modeling_wav2vec2.py:1094]:
```suggestion
            mask_feature_indices = torch.tensor(mask_feature_indices, device=hidden_states.device, dtype=torch.long)[:, None].expand(-1, sequence_length, -1)
```

File: examples/pytorch/speech-pretraining/README.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+	--saving_steps="10000" \
+	--per_device_train_batch_size="16" \
+	--per_device_eval_batch_size="16" \
+	--adam_beta1="0.9" \
+	--adam_beta2="0.98" \
+	--adam_epsilon="1e-06" \
+	--gradient_checkpointing \
+```
+
+The results of this run can be seen here: TODO
```
Comment (patrickvonplaten) [inline examples/pytorch/speech-pretraining/README.md:84]: Currently running this....

File: examples/pytorch/_tests_requirements.txt [hunk 4]
```diff
@@ -3,6 +3,7 @@ scikit-learn
 seqeval
 psutil
 sacrebleu >= 1.4.12
+accelerate >= 0.5.0
```
Comment (patrickvonplaten) [inline examples/pytorch/_tests_requirements.txt:6]: @sgugger - think we can start testing the `accelerate` scripts as well no?

File: examples/pytorch/speech-pretraining/run_wav2vec2_pretraining_no_trainer.py [hunk 5]
```diff
@@ -0,0 +1,701 @@
...
+                outputs.contrastive_loss.detach()
+                outputs.diversity_loss.detach()
+
+                if accelerator.state.num_processes > 1:
+                    dist.all_reduce(loss)
+                    dist.all_reduce(outputs.contrastive_loss)
+                    dist.all_reduce(outputs.diversity_loss)
+                    dist.all_reduce(percent_masked)
+
+                train_logs = {
```
Comment (patrickvonplaten) [inline examples/pytorch/speech-pretraining/run_wav2vec2_pretraining_no_trainer.py:621]: All the special logging, gradient averaging, etc...is easier to write down with `accelerate` IMO :-) `accelerate` works very nice for this use case here

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py [hunk 6]
```diff
@@ -115,19 +120,21 @@ class Wav2Vec2ForPreTrainingOutput(ModelOutput):
     codevector_perplexity: torch.FloatTensor = None
     hidden_states: Optional[Tuple[torch.FloatTensor]] = None
     attentions: Optional[Tuple[torch.FloatTensor]] = None
+    contrastive_loss: Optional[torch.FloatTensor] = None
+    diversity_loss: Optional[torch.FloatTensor] = None
 
 
 def _compute_mask_indices(
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/modeling_wav2vec2.py:127]: This function is changed in a breaking way, but it was always private and never used outside of the model for non-pretraining tasks, so that should be fine. 

Still have to verify that all the ASR downstream fine-tuning still gives same good results? Do we have to check `audio-classification` as well - does it make use of `mask_time_prob`? @anton-l

File: examples/pytorch/speech-pretraining/run_wav2vec2_pretraining_no_trainer.py [hunk 7]
```diff
@@ -0,0 +1,701 @@
...
+            # are accumulated for multiple backward passes in PyTorch
+            loss = outputs.loss / args.gradient_accumulation_steps
+            accelerator.backward(loss)
+
+            # make sure that `num_losses` is summed for distributed training
+            # and average gradients over losses of all devices
+            if accelerator.state.num_processes > 1:
+                dist.all_reduce(num_losses)
+                gradient_multiplier = accelerator.state.num_processes / num_losses
+                multiply_grads(model.module.parameters(), gradient_multiplier)
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply examples/pytorch/speech-pretraining/run_wav2vec2_pretraining_no_trainer.py:568]: yeah I still had it, but should probably test again (there were also some other problems before...) 

For DDP I wasn't sure whether `group_by_length` also makes sure that a batch that is distributed over multiple devices has examples of the same length or if it's just "grouped" per device - it should be "grouped" over all devices in DDP no?

[the remark this answers is not in the document]
Comment (stas00) [reply examples/pytorch/speech-pretraining/run_wav2vec2_pretraining_no_trainer.py:568]: > should it be easy to implement the gradient averaging in deepspeed as well? How should it be handled there?

I have never tried. Perhaps let's cross this bridge once this is being worked on in the Trainer?

I can meanwhile ask how to best approach that, but it'd help to know how this is going to be approached in the Trainer.

[the remark this answers is not in the document]
Comment (sgugger) [reply examples/pytorch/speech-pretraining/run_wav2vec2_pretraining_no_trainer.py:568]: > it should be "grouped" over all devices in DDP no?

Yes, it will be grouped over all devices :-)

**Review by sgugger — APPROVED** [review]
Nice new example! Thanks a lot for writing it!

**Comment by stas00** [comment]
question: why the other examples were not ported over? 

The deepspeed tests use `wav2vec2/run_asr.py` so I'm not quite sure how to do the transition. This is one of the most non-standard models so it'd be great to be able to continue running deepspeed tests.
